packages/vite/src/node/plugins/resolve.ts

```diff
@@ -142,6 +142,19 @@ export function resolvePlugin(resolveOptions: InternalResolveOptions): Plugin {
     name: 'vite:resolve',
 
     async resolveId(id, importer, resolveOpts) {
+      if (
+        id[0] === '\0' ||
+        id.startsWith('virtual:') ||
+        // When injected directly in html/client code
+        id.startsWith('/virtual:') ||
+        // This virtual file from the React plugin is hardcoded in integrations
+        // like Ruby & Laravel so adding this instead of migrating both React
+        // plugins to use /virtual:
+        id === '/@react-refresh'
```

Comment:
I'm not sure about hardcoding this, don't we already resolve this id in a pre plugin?

https://github.com/vitejs/vite-plugin-react/blob/a1b4de121b8baa75b21db35b8c77099f336a8bd2/packages/plugin-react/src/index.ts#L359-L363

That means it wouldn't hit here if it resolved there first.

Reply:
The SWC plugin is not pre. But maybe I can split the plugin in two so that the client is pre and the rest is not pre.
I think I will do the same because it's not easy to add pre-process of other plugin for example: https://github.com/vitejs/vite-plugin-react/pull/122/files?w=1#diff-885677e87d0c22a62efc8688d62685a305a3eb34469ddf0d2857da84cf417630R7